contracts/liquidity-protection/LiquidityProtection.sol

```diff
@@ -639,42 +632,147 @@ contract LiquidityProtection is ILiquidityProtection, Utils, Owned, ReentrancyGu
 
         // transfer the base tokens to the caller
         uint256 baseBalance = removedPos.reserveToken.balanceOf(address(this));
-        if (isMigrating) {
-            uint256 value;
-            if (removedPos.reserveToken.isNativeToken()) {
-                value = baseBalance;
-            } else {
-                IERC20(address(removedPos.reserveToken)).safeApprove(address(_networkV3), baseBalance);
+        removedPos.reserveToken.safeTransfer(provider, baseBalance);
+
+        // compensate the caller with network tokens if still needed
+        uint256 delta = _networkCompensation(targetAmount, baseBalance, packedRates);
+        if (delta > 0) {
+            // check if there's enough network token balance, otherwise mint more
+            uint256 networkBalance = _networkToken.balanceOf(address(this));
+            if (networkBalance < delta) {
+                _networkTokenGovernance.mint(address(this), delta - networkBalance);
             }
-            _networkV3.migrateLiquidity{ value: value }(
-                removedPos.reserveToken,
-                provider,
+
+            // lock network tokens for the caller
+            _networkToken.safeTransfer(address(_wallet), delta);
+            _lockTokens(provider, delta);
+        }
+
+        // if the contract still holds network tokens, burn them
+        uint256 networkBalance = _networkToken.balanceOf(address(this));
+        if (networkBalance > 0) {
+            _burnNetworkTokens(removedPos.poolToken, networkBalance);
+        }
+    }
+
+    /**
+     * @dev migrates a set of position lists to v3
+     *
+     * Requirements:
+     *
+     * - the caller must be the owner of all of the positions
+     */
+    function migratePositions(MigrationUnit[] calldata migrationUnits) external nonReentrant {
```

Comment:
I don't quite understand the flow - this is an external function while the migration flow needs only a list of position ids as input - doesn't make sense to me to expect the caller to input this new struct (and a good frontend is still not a solution).
the proper flow should be:
- receive the list of position ids
- build the input struct for the internal flow
- execute

I understand it will increase the contract size but we need to find a way to do that.

Reply:
What about the original suggestion to provide multiple arrays (which is similar to `MigrationUnit`, just omitting the pool and reserve token and fetching them from the first position)?

Reply:
it's better but 1 array is still cleaner - main issue is the binary size or something else?

Reply:
One array is both bad UX and also adds unnecessary logic to the code itself. Call?

Reply:
sure

Reply:
My original design was to fetch the pool token and the reserve token from the position of the first id, and then verify on the position of each one of the remaining IDs.
However, I then realized that it is better to just have the user (FE in this case) passing these tokens as part of the input, because of the following advantages:
1. Most importantly - it compels the user to "know what he's doing"
2. It reduces the gas cost of the transaction
3. It reduces the contract byte-code size
4. It simplifies the code

Reply:
discussed with @lbeder , array of arrays might be a bit better but it's not worth it as it'll require a frontend anyway, so might as well just leave it as is (with some renames - different discussion)